### ISSUE TITLE
about the class pred img

### ISSUE THREAD

**Issue by muzi2045** [kosuke55/train_baiducnn#8]
Hi, there has one problem about the class_pred img is not normal.
but the confidence img is ok.
![image](https://user-images.githubusercontent.com/12984898/82513325-3c9a1100-9b45-11ea-9143-d43c4691754a.png)

here is some data settting in the network:
width = 480
height = 480
grid_range = 48
class_num = 6 

thanks for any advice.
@kosuke55

**Comment by kosuke55** [comment]
@muzi2045 
Hi, thank you for your comment.
The correct prediction results look like this.
<img src="https://user-images.githubusercontent.com/39142679/82536985-498a2500-9b84-11ea-8c47-0a7cd2e2a4c7.png" width="500">
The weight of the black unknown class is lowered because the priority of the black unknown class is lower.  The unknown can be seen from the confidence.
By adjusting this class weight, you can also predict the unknown class. (There was a problem with the names of object_idx and nonobject_idx so I fixed it earlier.)
https://github.com/kosuke55/train_baiducnn/blob/dabc9a3237434526cbe97d95b89487de9e05c2d7/scripts/pytorch/train_bcnn.py#L88-L102
<img src="https://user-images.githubusercontent.com/39142679/82537062-6a527a80-9b84-11ea-9948-28191e973f78.png" width="500">

However, yours doesn't seem to predict the other classes correctly either.  
Did you get it right when width=672 range=70?
Adjusting the weight of the loss may help.
https://github.com/kosuke55/train_baiducnn/blob/dabc9a3237434526cbe97d95b89487de9e05c2d7/scripts/pytorch/train_bcnn.py#L109-L134

**Comment by muzi2045** [comment]
thanks for reply!
maybe is this part different from mine, I'll try it now.
```
class_weight = out_feature_gt.detach().numpy().copy() 
class_weight = class_weight[:, 4:5, ...] 
object_idx = np.where(class_weight == 0) 
nonobject_idx = np.where(class_weight != 0) 
class_weight[object_idx] = 1. 
class_weight[nonobject_idx] = 0.01 ** 2 
```

**Comment by muzi2045** [comment]
unforunately, the class img in my own dataset still get abnormal pred result.
And I tried to train this network in the nusc-mini dataset
the setting is wdith=672, height=672, range= 70, batch_size=8
here is the class pred img when start training, it looks like different from your post.
![image](https://user-images.githubusercontent.com/12984898/82623179-21490780-9c12-11ea-84ca-3171e25003dd.png)
And the code is same as the branch original_model in this repo.
it's really confused.
@kosuke55

**Comment by kosuke55** [comment]
@muzi2045 
The path of the trained model was incorrect and has been fixed. We apologize for the inconvenience. 
Please change the model name to `checkpoints/mini_672_6c.pt` like this
https://github.com/kosuke55/train_baiducnn/blob/acc5011153795ed6dd7bc1b4d73396caab3d185a/scripts/pytorch/train_bcnn.py#L46
https://github.com/kosuke55/train_baiducnn/blob/acc5011153795ed6dd7bc1b4d73396caab3d185a/scripts/pytorch/train_bcnn.py#L389

**Comment by kosuke55** [comment]
@muzi2045 
The path of the trained model was incorrect and has been fixed. I apologize for the inconvenience. 
Please change the model name to `checkpoints/mini_672_6c.pt` like this
https://github.com/kosuke55/train_baiducnn/blob/acc5011153795ed6dd7bc1b4d73396caab3d185a/scripts/pytorch/train_bcnn.py#L46
https://github.com/kosuke55/train_baiducnn/blob/acc5011153795ed6dd7bc1b4d73396caab3d185a/scripts/pytorch/train_bcnn.py#L389

**Comment by kosuke55** [comment]
@muzi2045 
The path of the trained model was incorrect and has been fixed. I apologize for the inconvenience. 
Please change the model name to `checkpoints/mini_672_6c.pt` like this
https://github.com/kosuke55/train_baiducnn/blob/acc5011153795ed6dd7bc1b4d73396caab3d185a/scripts/pytorch/train_bcnn.py#L46
https://github.com/kosuke55/train_baiducnn/blob/acc5011153795ed6dd7bc1b4d73396caab3d185a/scripts/pytorch/train_bcnn.py#L389
And even if you don't use my trained model, you should still be able to learn if you proceed with epoch. In my experience, there is a good learning order, and I feel that we can get better results if we let the model learn the confidence first.

**Comment by muzi2045** [comment]
thanks for your reply!
I have tried the pretrained model mini_672_6c.pt, and the class pred looks like no problem in nusc-mini dataset.
![image](https://user-images.githubusercontent.com/12984898/82637121-b231da00-9c36-11ea-97d7-e92f8c6ed08a.png)

But If I want to train with my own data from scratch(without the pretrained model), how could I produce the normal class pres img?
should I first learn the confidence pred ? is there any training log could be refering.
@kosuke55

**Comment by kosuke55** [comment]
@muzi2045 
Sorry, I don't have a detailed training log. My model first trained the confidence and then used it as a trained model to train the class. In order to learn well, the parameters of the optimaizer and loss have to be adjusted manually.
https://github.com/kosuke55/train_baiducnn/blob/dabc9a3237434526cbe97d95b89487de9e05c2d7/scripts/pytorch/train_bcnn.py#L71
If you want a normal class image, try setting the loss function to class only once, since the other instance_pt and height_pt trainings may be having a bad effect.
https://github.com/kosuke55/train_baiducnn/blob/dabc9a3237434526cbe97d95b89487de9e05c2d7/scripts/pytorch/train_bcnn.py#L109-L134

**Comment by muzi2045** [comment]
I found it really hard to train for correct cls pred img on my own data.
maybe the GT_class label is not correct? because the pedestrian class label are too small to predict. and the nusc-mini class label looks like Expand itself.

the follow image is only trained on confidence_loss and category loss, and the loss are going higher...

![image](https://user-images.githubusercontent.com/12984898/82728976-842bc300-9d26-11ea-9c5c-0f77040a5fc8.png)
@kosuke55

**Comment by kosuke55** [comment]
@muzi2045 
Yes, the pedestrian class is small, so the original_model branch annotates the label with a larger one. 
Your image class does not seem to be enlarged, but are you using data generated by the old create_dataset or my master branch?
Doesn't it work to train the class as a pretrained_model for a model that has been trained with confidence? 
If it doesn't work, what happens if you only train the class? It would be better to set the class weight of unknown to 1 to visualize it.

**Comment by muzi2045** [comment]
Hello, thanks for reply.
I an using data generated in the repo master branch, so the pedestrian and cyclist class label looks so small.
I have tried to trian those data on the master branch train code.
here is some result, still on trianing, but the class pred img looks like make sense now:
![image](https://user-images.githubusercontent.com/12984898/82966000-5aabb980-9ffc-11ea-94d7-485275017201.png)
![image](https://user-images.githubusercontent.com/12984898/82966013-61d2c780-9ffc-11ea-8b5c-3c12fed5b316.png)
![image](https://user-images.githubusercontent.com/12984898/82966048-7616c480-9ffc-11ea-964e-592ada1e0bc3.png)

@kosuke55
